File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = ""
        closes = [')', ']', '}']
        opens = ['(', '[', '{']
        for ss in s:
            if not stack:
                stack += ss
            else:
                if (ss == closes[0] and stack[-1] == opens[0]) or (ss == closes[1] and stack[-1] == opens[1]) or (ss == closes[2] and stack[-1] == opens[2]):
                    stack = stack[:-1]
                else:
                    stack += ss
        print(stack)
        return True if not stack else False

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack();
        for (int i = 0; i < s.length(); i++) {
            char current = s.charAt(i);
            if (current == '(' || current == '[' || current == '{') {
                stack.push(current);
            } else {
                if (stack.isEmpty()) {
                    return false;
                }
                char last = stack.pop();
                if ((current == ')' && last != '(') ||
                    (current == ']' && last != '[') ||
                    (current == '}' && last != '{')) {
                    return false;
                }
            }
        }

        if (stack.isEmpty()) {
            return true;
        }
        return false;
    }
    
}